File: 34searchRange.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        def find_left_index(nums, target):
            low, high = 0, len(nums) - 1
            while low < high:
                mid = low + (high - low) // 2
                if nums[mid] > target or target == nums[mid]:
                    high = mid  # 找左边界的时候，由右侧逼近。
                else:
                    low = mid + 1
            return low if nums and nums[low] == target else -1

        def find_right_index(nums, target):
            low, high = 0, len(nums) - 1
            while low < high:
                mid = low + (high - low) // 2 + 1
                if nums[mid] < target or target == nums[mid]:
                    low = mid  # 找右边界的时候，由左侧逼近
                else:
                    high = mid - 1
            return low if nums and nums[low] == target else -1

        return [find_left_index(nums, target), find_right_index(nums, target)]

File: Leetcode_with_Cpp/34searchRange.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        if(nums.empty())
            return {-1, -1};
		int l = 0, r = nums.size() - 1;
		while(l < r) {  //找到起始位置 
			int mid = l+r >> 1;
			if(nums[mid] >= target)
                r = mid;
			else l = mid + 1;
		}
		if (nums[l] != target) 
            return {-1,-1};
		int start = l;
		r = nums.size() - 1;
		while(l < r) {  //找到终点位置
			int mid = l+r+1 >> 1;  // 不加括号效率更高
			if (nums[mid] > target)
                r = mid - 1;
			else l = mid;
		}
		return {start,r};
    }
};